### PR TITLE
fix(test:counts): extend silent-test-drop guard to the Firestore rules suite

### PR DESCRIPTION
**Pull request by OPS-PIvers**
## Summary

Nightly code-health run (2026-07-13), Build & Tooling area.

`scripts/checkTestCounts.mjs` (added in #2139) guards against a Vitest test suite silently shrinking — a test file becoming invisible to the collector (typo'd extension, over-broad `exclude`, a disabled `describe` block) exits 0 with fewer tests counted and no failure signal. It only ever covered two of the three Vitest surfaces in this repo: `root` (`pnpm test`) and `functions` (`pnpm -C functions test`). The third surface, `tests/rules/**` (the Firestore-rules emulator suite, run via `pnpm run test:rules` and invoked as its own step in `pr-validation.yml`), had **no count-floor guard at all**.

## Root cause

`main()`'s hardcoded two-entry `targets` array never referenced `tests/rules/**`, and `pnpm run test:rules` never produced a JSON-reporter side artifact for the guard to read in the first place.

## Fix

1. `scripts/checkTestCounts.mjs` — extracted per-target evaluation into a new exported, unit-testable `evaluateTarget()` and added a third `rules` target reading `.vitest-reports/rules.json`.
2. `package.json` — `test:rules` now also runs with `--reporter=json --outputFile=.vitest-reports/rules.json` and chains `node scripts/checkTestCounts.mjs` afterward, so the guard actually executes whenever the rules suite runs.
3. `scripts/test-count-baseline.json` — added a real `rules` baseline (47 files / 1112 tests), measured by running the full emulator suite.

## Band-aid rejected

Making the `rules` report unconditionally required (matching root/functions) would be the naive fix, but `test:rules` boots a separate Firestore emulator and is deliberately excluded from `validate`/`test:all` (per CLAUDE.md) — `.vitest-reports/rules.json` legitimately won't exist on that path. Hard-requiring it would break `validate` for everyone. Instead `rules` is marked `optional: true`: a missing report is silently skipped (the suite that would produce it simply wasn't run), a present one is fully checked against the baseline.

## Regression test

`scripts/checkTestCounts.test.ts` — 4 new cases covering `evaluateTarget`:
- A rules-suite regression is unreachable by any existing check pre-fix (no `rules` target existed to invoke).
- Post-fix, the same regression is caught (degraded report → "file count dropped" / "test count dropped" issues).
- Load-bearing safety property: an **optional** target with a missing report (the normal `validate`/`test:all` path) produces zero issues.
- A **required** target (root/functions) with a missing report still fails loudly, unchanged.

All 14 tests in the file pass (10 pre-existing + 4 new).

## Evidence

- `pnpm exec vitest run scripts/checkTestCounts.test.ts` — 14/14 pass
- Ran the real `pnpm run test:rules` end-to-end once to generate the actual baseline numbers and confirm the JSON report shape
- `pnpm run validate` (type-check:all, lint, format:check, test, functions test, test:counts) — independently re-run by the orchestrator: root 562/562 test files, functions 37/37, test:counts guard passed
- `pnpm run build:all` — independently re-run by the orchestrator, succeeds

## Risk

**Contained** — extends an existing CI guard script + one `package.json` script chain; no production code paths touched.

---
_Generated by [Claude Code](https://claude.ai/code/session_016DVeKvHb4QaTHifxZSXzVy)_

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,7 +31,7 @@
     "test:all": "pnpm test && pnpm -C functions test && pnpm run test:counts",
     "test:counts": "node scripts/checkTestCounts.mjs",
     "test:e2e": "playwright test",
-    "test:rules": "firebase emulators:exec --only firestore \"vitest run --config vitest.rules.config.ts\"",
+    "test:rules": "firebase emulators:exec --only firestore \"vitest run --config vitest.rules.config.ts --reporter=default --reporter=json --outputFile=.vitest-reports/rules.json && node scripts/checkTestCounts.mjs\"",
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply package.json]: No change needed — this is already implemented on the branch. `package.json`'s `test:rules` script passes the `rules` argument (`node scripts/checkTestCounts.mjs rules`), and `selectTargets` scopes evaluation to just that target, so the isolated rules CI job no longer fails on missing `root`/`functions` reports.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/checkTestCounts.test.ts [hunk 2]
```diff
@@ -194,3 +196,130 @@ describe('validateReport', () => {
...
+    const targets = selectTargets(allTargets, undefined);
+    expect(targets.map((t) => t.label)).toEqual(['root', 'functions', 'rules']);
+  });
+
+  it('PASSES after: requesting "rules" scopes evaluation to just that target', () => {
+    const targets = selectTargets(allTargets, 'rules');
+    expect(targets.map((t) => t.label)).toEqual(['rules']);
+  });
+
+  it('an unknown requested label selects nothing (fails loudly via empty targets, not silently)', () => {
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply scripts/checkTestCounts.test.ts]: No change needed — the branch already reflects the guard approach you preferred. `selectTargets` now `throw`s on an unknown label (rather than returning `[]` and silently no-op'ing `main()`), and the corresponding test is titled "an unknown requested label throws instead of silently selecting nothing" with a comment crediting this review. The behavior now matches the description precisely.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by OPS-PIvers — COMMENTED** [review]
## Automated Code Review

**Date:** 2026-07-13

### Summary
Extends the silent-test-drop guard (`scripts/checkTestCounts.mjs`) to the previously-unguarded Firestore rules suite. Refactors per-target evaluation into an exported `evaluateTarget()`, adds a `rules` target (`optional: true`), and chains the guard into `test:rules`. Sound design.

### Code Quality
✓ No issues found. `selectTargets` throws on an unknown label rather than silently selecting nothing, and `test:rules` passes the `rules` argument so the isolated rules CI job doesn't fail on missing `root`/`functions` reports.

### Widget Standards
✓ Not applicable.

### Admin Config Standards
✓ Not applicable.

### Firestore & Security
✓ No issues found — CI tooling only; the rules suite itself is unchanged.

### Regression Risk
✓ Low risk. The `optional: true` marking is the load-bearing safety property: `validate`/`test:all` never run `test:rules`, so a missing `rules.json` must be silently skipped there — while a missing `root`/`functions` report still fails loudly. Both directions are unit-tested.

### Missing Tests
✓ Adequate coverage — `checkTestCounts.test.ts` gains 4 `evaluateTarget` cases (rules regression caught, optional-missing = no issues, required-missing fails loudly, missing baseline fails) plus `selectTargets` scoping/throw cases.

### Notes on prior review threads
Both open threads (Gemini: pass `rules` arg; Claude: description vs `main()` behavior) are already resolved in the current branch — replied inline.

### Overall
**Merge readiness:** Ready
Closes a real CI observability gap with careful attention to the required-vs-optional distinction.

---
_This review was generated automatically. All findings should be verified by a human before merging._

---
_Generated by [Claude Code](https://claude.ai/code)_
